#include "warehousewindow.h"
#include "ui_warehousewindow.h"


WarehouseWindow::WarehouseWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::WarehouseWindow){
    ui->setupUi(this);

    ui->tableWidget->setColumnCount(4);
    ui->tableWidget->setHorizontalHeaderLabels({"–ù–∞–∑–≤–∞", "–í–∏—Ä–æ–±–Ω–∏–∫", "–¶—ñ–Ω–∞ (–≥—Ä–Ω)", "–ö—ñ–ª—å–∫—ñ—Å—Ç—å"});
    ui->tableWidget->horizontalHeader()->setStretchLastSection(true);
    ui->tableWidget->setSelectionBehavior(QAbstractItemView::SelectRows);
    ui->tableWidget->setEditTriggers(QAbstractItemView::NoEditTriggers);

    connect(ui->tableWidget, &QTableWidget::cellDoubleClicked, this, &WarehouseWindow::onTableCellDoubleClicked);

    addSampleParts();
    displayParts(warehouse.getAllParts());
    updateStats();
}

WarehouseWindow::~WarehouseWindow(){
    delete ui;
}

void WarehouseWindow::on_searchButton_clicked(){
    string manufacturer = ui->searchLineEdit->text().trimmed().toStdString();

    if (manufacturer.empty()){
        QMessageBox::warning(this, "–ü–æ–º–∏–ª–∫–∞", "–í–≤–µ–¥—ñ—Ç—å –≤–∏—Ä–æ–±–Ω–∏–∫–∞ –¥–ª—è –ø–æ—à—É–∫—É!");
        return;
    }

    vector<AutoPart> results = warehouse.findByManufacturer(manufacturer);

    if (results.empty()){
        QMessageBox::information(this, "–†–µ–∑—É–ª—å—Ç–∞—Ç–∏ –ø–æ—à—É–∫—É",
                                 QString("–ó–∞–ø—á–∞—Å—Ç–∏–Ω –≤—ñ–¥ –≤–∏—Ä–æ–±–Ω–∏–∫–∞ '%1' –Ω–µ –∑–Ω–∞–π–¥–µ–Ω–æ.")
                                 .arg(QString::fromStdString(manufacturer)));
    }

    displayParts(results);
}

void WarehouseWindow::on_showAllButton_clicked(){
    displayParts(warehouse.getAllParts());
}

void WarehouseWindow::on_sortNameButton_clicked(){
    vector<AutoPart> parts = warehouse.getAllParts();
    warehouse.sortByName(parts);

    if (!nameSortAscending){
        reverse(parts.begin(), parts.end());
    }

    displayParts(parts);
    nameSortAscending = !nameSortAscending;
}

void WarehouseWindow::on_sortPriceButton_clicked(){
    vector<AutoPart> parts = warehouse.getAllParts();
    warehouse.sortByPrice(parts);

    if (!priceSortAscending){
        reverse(parts.begin(), parts.end());
    }

    displayParts(parts);
    priceSortAscending = !priceSortAscending;
}

// ========== –ù–û–í–Ü –§–£–ù–ö–¶–Ü–á LAB 8 ==========

void WarehouseWindow::on_addSupplyButton_clicked(){
    QDialog dialog(this);
    dialog.setWindowTitle("–î–æ–¥–∞—Ç–∏ –ø–æ—Å—Ç–∞—á–∞–Ω–Ω—è (QUEUE)");
    dialog.setMinimumWidth(450);

    QVBoxLayout *mainLayout = new QVBoxLayout(&dialog);
    QFormLayout *formLayout = new QFormLayout();

    QLineEdit *nameEdit = new QLineEdit();
    QLineEdit *manufacturerEdit = new QLineEdit();
    QDoubleSpinBox *priceSpinBox = new QDoubleSpinBox();
    priceSpinBox->setRange(0, 1000000);
    priceSpinBox->setDecimals(2);
    priceSpinBox->setSuffix(" –≥—Ä–Ω");

    QSpinBox *quantitySpinBox = new QSpinBox();
    quantitySpinBox->setRange(1, 10000);
    quantitySpinBox->setValue(10);

    QCheckBox *urgentCheckBox = new QCheckBox("–¢–µ—Ä–º—ñ–Ω–æ–≤–µ –ø–æ—Å—Ç–∞—á–∞–Ω–Ω—è");

    formLayout->addRow("–ù–∞–∑–≤–∞:", nameEdit);
    formLayout->addRow("–í–∏—Ä–æ–±–Ω–∏–∫:", manufacturerEdit);
    formLayout->addRow("–¶—ñ–Ω–∞:", priceSpinBox);
    formLayout->addRow("–ö—ñ–ª—å–∫—ñ—Å—Ç—å:", quantitySpinBox);
    formLayout->addRow("", urgentCheckBox);

    mainLayout->addLayout(formLayout);

    QDialogButtonBox *buttonBox = new QDialogButtonBox(
        QDialogButtonBox::Ok | QDialogButtonBox::Cancel);
    
    QPushButton *cancelBtn = buttonBox->button(QDialogButtonBox::Cancel);
    if (cancelBtn) cancelBtn->setProperty("class", "secondary");
    
    connect(buttonBox, &QDialogButtonBox::accepted, &dialog, &QDialog::accept);
    connect(buttonBox, &QDialogButtonBox::rejected, &dialog, &QDialog::reject);
    mainLayout->addWidget(buttonBox);

    if (dialog.exec() == QDialog::Accepted){
        QString name = nameEdit->text().trimmed();
        QString manufacturer = manufacturerEdit->text().trimmed();

        if (name.isEmpty() || manufacturer.isEmpty()){
            QMessageBox::warning(this, "–ü–æ–º–∏–ª–∫–∞", "–ó–∞–ø–æ–≤–Ω—ñ—Ç—å –≤—Å—ñ –ø–æ–ª—è!");
            return;
        }

        AutoPart part(name.toStdString(), manufacturer.toStdString(),
                      priceSpinBox->value(), quantitySpinBox->value());

        warehouse.addSupplyOrder(part, urgentCheckBox->isChecked());
        updateStats();

        QMessageBox::information(this, "–£—Å–ø—ñ—Ö",
                                 QString("–ü–æ—Å—Ç–∞—á–∞–Ω–Ω—è –¥–æ–¥–∞–Ω–æ –≤ —á–µ—Ä–≥—É!\n–í—Å—å–æ–≥–æ –≤ —á–µ—Ä–∑—ñ: %1")
                                     .arg(warehouse.getSupplyQueueSize()));
    }
}

void WarehouseWindow::on_processSupplyButton_clicked(){
    if (warehouse.getSupplyQueueSize() == 0){
        QMessageBox::information(this, "–Ü–Ω—Ñ–æ—Ä–º–∞—Ü—ñ—è", "–ß–µ—Ä–≥–∞ –ø–æ—Å—Ç–∞—á–∞–Ω—å –ø–æ—Ä–æ–∂–Ω—è!");
        return;
    }

    if (warehouse.processNextSupply()){
        updateStats();
        on_showAllButton_clicked();
        QMessageBox::information(this, "–£—Å–ø—ñ—Ö",
                                 QString("–ü–æ—Å—Ç–∞—á–∞–Ω–Ω—è –æ–±—Ä–æ–±–ª–µ–Ω–æ!\n–ó–∞–ª–∏—à–∏–ª–æ—Å—å –≤ —á–µ—Ä–∑—ñ: %1")
                                     .arg(warehouse.getSupplyQueueSize()));
    } else {
        QMessageBox::warning(this, "–ü–æ–º–∏–ª–∫–∞", "–ù–µ –≤–¥–∞–ª–æ—Å—è –æ–±—Ä–æ–±–∏—Ç–∏ –ø–æ—Å—Ç–∞—á–∞–Ω–Ω—è!");
    }
}

void WarehouseWindow::on_showUniqueNamesButton_clicked(){
    auto uniqueNames = warehouse.getUniqueNames();

    QString message = "<b>SET - –£–Ω—ñ–∫–∞–ª—å–Ω—ñ –Ω–∞–∑–≤–∏ –∑–∞–ø—á–∞—Å—Ç–∏–Ω:</b><br><br>";
    message += QString("–í—Å—å–æ–≥–æ —É–Ω—ñ–∫–∞–ª—å–Ω–∏—Ö –Ω–∞–∑–≤: %1<br><br>").arg(uniqueNames.size());

    int count = 1;
    for (const auto& name : uniqueNames){
        message += QString("%1. %2<br>").arg(count++).arg(QString::fromStdString(name));
    }

    QMessageBox msgBox(this);
    msgBox.setWindowTitle("–£–Ω—ñ–∫–∞–ª—å–Ω—ñ –Ω–∞–∑–≤–∏ (SET)");
    msgBox.setTextFormat(Qt::RichText);
    msgBox.setText(message);
    msgBox.setStandardButtons(QMessageBox::Ok);
    msgBox.exec();
}

void WarehouseWindow::on_showSupplyQueueButton_clicked() {
    auto orders = warehouse.getAllSupplyOrders();

    if (orders.empty()) {
        QMessageBox::information(this, "–ß–µ—Ä–≥–∞ –ø–æ—Å—Ç–∞—á–∞–Ω—å", "–ß–µ—Ä–≥–∞ –ø–æ—Ä–æ–∂–Ω—è!");
        return;
    }

    QString message = "<b>QUEUE - –ß–µ—Ä–≥–∞ –ø–æ—Å—Ç–∞—á–∞–Ω—å:</b><br><br>";
    message += QString("–í—Å—å–æ–≥–æ –≤ —á–µ—Ä–∑—ñ: %1<br><br>").arg(orders.size());

    int count = 1;
    for (const auto& order : orders) {
        message += QString("<b>%1.</b> %2 (%3)<br>")
        .arg(count++)
            .arg(QString::fromStdString(order.part.getName()))
            .arg(QString::fromStdString(order.part.getManufacturer()));
        message += QString("   –ö—ñ–ª—å–∫—ñ—Å—Ç—å: %1, –î–∞—Ç–∞: %2%3<br>")
                       .arg(order.part.getQuantity())
                       .arg(QString::fromStdString(order.orderDate))
                       .arg(order.isUrgent ? " <b>[–¢–ï–†–ú–Ü–ù–û–í–ï]</b>" : "");
    }

    QMessageBox msgBox(this);
    msgBox.setWindowTitle("–ß–µ—Ä–≥–∞ –ø–æ—Å—Ç–∞—á–∞–Ω—å (QUEUE)");
    msgBox.setTextFormat(Qt::RichText);
    msgBox.setText(message);
    msgBox.setStandardButtons(QMessageBox::Ok);
    msgBox.exec();
}

void WarehouseWindow::on_showLowStockButton_clicked(){
    auto lowStock = warehouse.getLowStockParts(10);

    if (lowStock.empty()){
        QMessageBox::information(this, "–ù–∏–∑—å–∫—ñ –∑–∞–ø–∞—Å–∏", "–ù–µ–º–∞—î –∑–∞–ø—á–∞—Å—Ç–∏–Ω –Ω–∞ —Å–∫–ª–∞–¥—ñ!");
        return;
    }

    QString message = "<b>PRIORITY_QUEUE - –ó–∞–ø—á–∞—Å—Ç–∏–Ω–∏ –∑ –Ω–∞–π–º–µ–Ω—à–∏–º–∏ –∑–∞–ø–∞—Å–∞–º–∏:</b><br><br>";
    message += "–†–µ–∫–æ–º–µ–Ω–¥–æ–≤–∞–Ω–æ –ø—Ä—ñ–æ—Ä–∏—Ç–µ—Ç–Ω–µ –ø–æ–ø–æ–≤–Ω–µ–Ω–Ω—è:<br><br>";

    int count = 1;
    for (const auto& part : lowStock){
        QString priority = "";
        if (part.getQuantity() < 5) priority = " <span style='color:red;'>[–ö–†–ò–¢–ò–ß–ù–û]</span>";
        else if (part.getQuantity() < 10) priority = " <span style='color:orange;'>[–ù–ò–ó–¨–ö–û]</span>";

        message += QString("<b>%1.</b> %2 (%3)%4<br>")
                       .arg(count++)
                       .arg(QString::fromStdString(part.getName()))
                       .arg(QString::fromStdString(part.getManufacturer()))
                       .arg(priority);
        message += QString("   –ó–∞–ª–∏—à–æ–∫: <b>%1 —à—Ç.</b>, –¶—ñ–Ω–∞: %2 –≥—Ä–Ω<br>")
                       .arg(part.getQuantity())
                       .arg(part.getPrice(), 0, 'f', 2);
    }

    QMessageBox msgBox(this);
    msgBox.setWindowTitle("–ù–∏–∑—å–∫—ñ –∑–∞–ø–∞—Å–∏ (PRIORITY_QUEUE)");
    msgBox.setTextFormat(Qt::RichText);
    msgBox.setText(message);
    msgBox.setStandardButtons(QMessageBox::Ok);
    msgBox.exec();
}

void WarehouseWindow::on_showHistoryButton_clicked(){
    if (!warehouse.hasHistory()){
        QMessageBox::information(this, "–Ü—Å—Ç–æ—Ä—ñ—è", "–Ü—Å—Ç–æ—Ä—ñ—è –∑–º—ñ–Ω –ø–æ—Ä–æ–∂–Ω—è!");
        return;
    }

    auto history = warehouse.getHistory(20);

    QString message = "<b>STACK - –Ü—Å—Ç–æ—Ä—ñ—è –∑–º—ñ–Ω (–æ—Å—Ç–∞–Ω–Ω—ñ –¥—ñ—ó):</b><br><br>";
    message += QString("–í—Å—å–æ–≥–æ –∑–∞–ø–∏—Å—ñ–≤: –ø–æ–∫–∞–∑–∞–Ω–æ –æ—Å—Ç–∞–Ω–Ω—ñ %1<br><br>").arg(history.size());

    int count = 1;
    for (const auto& entry : history){
        QString actionColor;
        if (entry.action == "ADD") actionColor = "green";
        else if (entry.action == "DELETE") actionColor = "red";
        else if (entry.action == "UPDATE") actionColor = "blue";
        else actionColor = "orange";

        message += QString("<b>%1.</b> <span style='color:%2;'>[%3]</span> %4 (%5)<br>")
                       .arg(count++)
                       .arg(actionColor)
                       .arg(QString::fromStdString(entry.action))
                       .arg(QString::fromStdString(entry.part.getName()))
                       .arg(QString::fromStdString(entry.part.getManufacturer()));
        message += QString("   –ß–∞—Å: %1, –¶—ñ–Ω–∞: %2 –≥—Ä–Ω, –ö—ñ–ª—å–∫—ñ—Å—Ç—å: %3<br>")
                       .arg(QString::fromStdString(entry.timestamp))
                       .arg(entry.part.getPrice(), 0, 'f', 2)
                       .arg(entry.part.getQuantity());
    }

    QMessageBox msgBox(this);
    msgBox.setWindowTitle("–Ü—Å—Ç–æ—Ä—ñ—è –∑–º—ñ–Ω (STACK)");
    msgBox.setTextFormat(Qt::RichText);
    msgBox.setText(message);
    msgBox.setStandardButtons(QMessageBox::Ok);
    msgBox.exec();
}

void WarehouseWindow::on_showMultimapButton_clicked(){
    QDialog dialog(this);
    dialog.setWindowTitle("–ü–æ—à—É–∫ –¥—É–±–ª—å–æ–≤–∞–Ω–∏—Ö –∑–∞–ø—á–∞—Å—Ç–∏–Ω (MULTIMAP)");
    dialog.setMinimumWidth(400);

    QVBoxLayout *layout = new QVBoxLayout(&dialog);

    QLabel *label = new QLabel("–í–≤–µ–¥—ñ—Ç—å –Ω–∞–∑–≤—É –∑–∞–ø—á–∞—Å—Ç–∏–Ω–∏ –¥–ª—è –ø–æ—à—É–∫—É –≤—Å—ñ—Ö –≤–∞—Ä—ñ–∞–Ω—Ç—ñ–≤:");
    layout->addWidget(label);

    QLineEdit *searchEdit = new QLineEdit();
    layout->addWidget(searchEdit);

    QTextEdit *resultText = new QTextEdit();
    resultText->setReadOnly(true);
    layout->addWidget(resultText);

    QPushButton *searchBtn = new QPushButton("–ü–æ—à—É–∫");
    layout->addWidget(searchBtn);

    connect(searchBtn, &QPushButton::clicked, [&](){
        QString name = searchEdit->text().trimmed();
        if (name.isEmpty()){
            resultText->setHtml("<span style='color:red;'>–í–≤–µ–¥—ñ—Ç—å –Ω–∞–∑–≤—É!</span>");
            return;
        }

        auto parts = warehouse.findAllByName(name.toStdString());

        if (parts.empty()){
            resultText->setHtml(QString("<span style='color:orange;'>–ó–∞–ø—á–∞—Å—Ç–∏–Ω –∑ –Ω–∞–∑–≤–æ—é '%1' –Ω–µ –∑–Ω–∞–π–¥–µ–Ω–æ!</span>")
                                    .arg(name));
            return;
        }

        QString html = QString("<b>MULTIMAP - –ó–Ω–∞–π–¥–µ–Ω–æ %1 –≤–∞—Ä—ñ–∞–Ω—Ç(—ñ–≤) '%2':</b><br><br>")
                           .arg(parts.size())
                           .arg(name);

        int count = 1;
        for (const auto& part : parts){
            html += QString("<b>%1.</b> –í–∏—Ä–æ–±–Ω–∏–∫: %2<br>")
                        .arg(count++)
                        .arg(QString::fromStdString(part.getManufacturer()));
            html += QString("   –¶—ñ–Ω–∞: %1 –≥—Ä–Ω, –ö—ñ–ª—å–∫—ñ—Å—Ç—å: %2 —à—Ç.<br><br>")
                        .arg(part.getPrice(), 0, 'f', 2)
                        .arg(part.getQuantity());
        }

        resultText->setHtml(html);
    });

    QDialogButtonBox *buttonBox = new QDialogButtonBox(QDialogButtonBox::Close);
    QPushButton *closeBtn = buttonBox->button(QDialogButtonBox::Close);
    if (closeBtn) closeBtn->setProperty("class", "secondary");
    
    connect(buttonBox, &QDialogButtonBox::rejected, &dialog, &QDialog::reject);
    layout->addWidget(buttonBox);

    dialog.exec();
}

void WarehouseWindow::removePart(string name){
    QMessageBox::StandardButton reply;
    reply = QMessageBox::question(this, "–ü—ñ–¥—Ç–≤–µ—Ä–¥–∂–µ–Ω–Ω—è",
                                  QString("–í–∏–¥–∞–ª–∏—Ç–∏ –∑–∞–ø—á–∞—Å—Ç–∏–Ω—É '%1'?")
                                      .arg(QString::fromStdString(name)),
                                  QMessageBox::Yes | QMessageBox::No);

    if (reply == QMessageBox::Yes){
        if (warehouse.removePart(name)){
            updateStats();
            on_showAllButton_clicked();
            QMessageBox::information(this, "–£—Å–ø—ñ—Ö", "–ó–∞–ø—á–∞—Å—Ç–∏–Ω—É –≤–∏–¥–∞–ª–µ–Ω–æ!");
        } else {
            QMessageBox::warning(this, "–ü–æ–º–∏–ª–∫–∞", "–ù–µ –≤–¥–∞–ª–æ—Å—è –≤–∏–¥–∞–ª–∏—Ç–∏ –∑–∞–ø—á–∞—Å—Ç–∏–Ω—É!");
        }
    }
}

void WarehouseWindow::updatePart(string name){
    AutoPart* part = warehouse.findPart(name);

    if (!part){
        QMessageBox::warning(this, "–ü–æ–º–∏–ª–∫–∞", "–ó–∞–ø—á–∞—Å—Ç–∏–Ω—É –Ω–µ –∑–Ω–∞–π–¥–µ–Ω–æ!");
        return;
    }

    QDialog dialog(this);
    dialog.setWindowTitle("–û–Ω–æ–≤–ª–µ–Ω–Ω—è –∑–∞–ø—á–∞—Å—Ç–∏–Ω–∏");
    dialog.setMinimumWidth(400);

    QVBoxLayout *mainLayout = new QVBoxLayout(&dialog);

    QLabel *titleLabel = new QLabel(QString("–û–Ω–æ–≤–ª–µ–Ω–Ω—è: %1")
                                        .arg(QString::fromStdString(name)));
    QFont titleFont = titleLabel->font();
    titleFont.setPointSize(12);
    titleFont.setBold(true);
    titleLabel->setFont(titleFont);
    mainLayout->addWidget(titleLabel);

    QFormLayout *formLayout = new QFormLayout();

    QDoubleSpinBox *priceSpinBox = new QDoubleSpinBox();
    priceSpinBox->setRange(0, 1000000);
    priceSpinBox->setDecimals(2);
    priceSpinBox->setSuffix(" –≥—Ä–Ω");
    priceSpinBox->setValue(part->getPrice());

    QSpinBox *quantitySpinBox = new QSpinBox();
    quantitySpinBox->setRange(0, 10000);
    quantitySpinBox->setValue(part->getQuantity());

    formLayout->addRow("–¶—ñ–Ω–∞:", priceSpinBox);
    formLayout->addRow("–ö—ñ–ª—å–∫—ñ—Å—Ç—å:", quantitySpinBox);

    mainLayout->addLayout(formLayout);

    QDialogButtonBox *buttonBox = new QDialogButtonBox();
    QPushButton *updateButton = buttonBox->addButton("–û–Ω–æ–≤–∏—Ç–∏", QDialogButtonBox::AcceptRole);
    QPushButton *deleteButton = buttonBox->addButton("–í–∏–¥–∞–ª–∏—Ç–∏", QDialogButtonBox::DestructiveRole);
    deleteButton->setProperty("class", "destructive");
    QPushButton *cancelButton = buttonBox->addButton("–°–∫–∞—Å—É–≤–∞—Ç–∏", QDialogButtonBox::RejectRole);
    cancelButton->setProperty("class", "secondary");
    
    connect(updateButton, &QPushButton::clicked, &dialog, &QDialog::accept);
    connect(cancelButton, &QPushButton::clicked, &dialog, &QDialog::reject);
    connect(deleteButton, &QPushButton::clicked, [&](){
        dialog.done(2);
    });
    
    mainLayout->addWidget(buttonBox);

    int result = dialog.exec();
    
    if (result == QDialog::Accepted){
        warehouse.updatePart(name, priceSpinBox->value(), quantitySpinBox->value());
        updateStats();
        on_showAllButton_clicked();
        QMessageBox::information(this, "–£—Å–ø—ñ—Ö", "–ó–∞–ø—á–∞—Å—Ç–∏–Ω—É –æ–Ω–æ–≤–ª–µ–Ω–æ!");
    } else if (result == 2){
        removePart(name);
    }
}

void WarehouseWindow::displayParts(vector<AutoPart> parts){
    ui->tableWidget->setRowCount(0);

    for (size_t i = 0; i < parts.size(); ++i){
        const AutoPart& part = parts[i];

        int row = ui->tableWidget->rowCount();
        ui->tableWidget->insertRow(row);

        ui->tableWidget->setItem(row, 0,
                                 new QTableWidgetItem(QString::fromStdString(part.getName())));
        ui->tableWidget->setItem(row, 1,
                                 new QTableWidgetItem(QString::fromStdString(part.getManufacturer())));
        ui->tableWidget->setItem(row, 2,
                                 new QTableWidgetItem(QString::number(part.getPrice(), 'f', 2)));
        ui->tableWidget->setItem(row, 3,
                                 new QTableWidgetItem(QString::number(part.getQuantity())));

        QPushButton *updateButton = new QPushButton("–û–Ω–æ–≤–∏—Ç–∏");
        updateButton->setMaximumWidth(90);
        updateButton->setMinimumHeight(36);
        QPushButton *deleteButton = new QPushButton("–í–∏–¥–∞–ª–∏—Ç–∏");
        deleteButton->setMaximumWidth(90);
        deleteButton->setMinimumHeight(36);

        string key = part.getName() + "_" + part.getManufacturer();
        connect(updateButton, &QPushButton::clicked,
                [this, key](){ updatePart(key); });
        connect(deleteButton, &QPushButton::clicked,
                [this, key](){ removePart(key); });

        QWidget *widget = new QWidget();
        QHBoxLayout *layout = new QHBoxLayout(widget);

        layout->addWidget(updateButton);
        layout->addWidget(deleteButton);
        layout->setAlignment(Qt::AlignCenter);
        layout->setContentsMargins(5, 2, 5, 2);

        ui->tableWidget->setCellWidget(row, 4, widget);
        ui->tableWidget->setRowHeight(row, 50);
    }

    for (int i = 0; i < 4; ++i){
        ui->tableWidget->resizeColumnToContents(i);
    }
}

void WarehouseWindow::addSampleParts(){
    warehouse.addPart(AutoPart("–ú–∞—Å–ª—è–Ω–∏–π —Ñ—ñ–ª—å—Ç—Ä", "Bosch", 250.00, 3));
    warehouse.addPart(AutoPart("–ú–∞—Å–ª—è–Ω–∏–π —Ñ—ñ–ª—å—Ç—Ä", "Mann", 220.00, 15));
    warehouse.addPart(AutoPart("–°–≤—ñ—á–∫–∏ –∑–∞–ø–∞–ª—é–≤–∞–Ω–Ω—è", "NGK", 120.00, 2));
    warehouse.addPart(AutoPart("–ì–∞–ª—å–º—ñ–≤–Ω—ñ –∫–æ–ª–æ–¥–∫–∏", "Brembo", 1800.00, 10));
    warehouse.addPart(AutoPart("–ü–æ–≤—ñ—Ç—Ä—è–Ω–∏–π —Ñ—ñ–ª—å—Ç—Ä", "Mann", 350.00, 20));
    warehouse.addPart(AutoPart("–ê–∫—É–º—É–ª—è—Ç–æ—Ä", "Varta", 3500.00, 1));
    warehouse.addPart(AutoPart("–†–µ–º—ñ–Ω—å –ì–†–ú", "Gates", 890.00, 12));
    warehouse.addPart(AutoPart("–ê–º–æ—Ä—Ç–∏–∑–∞—Ç–æ—Ä", "Bilstein", 2200.00, 8));

    // –î–æ–¥–∞—î–º–æ –∫—ñ–ª—å–∫–∞ –ø–æ—Å—Ç–∞—á–∞–Ω—å –≤ —á–µ—Ä–≥—É
    warehouse.addSupplyOrder(AutoPart("–°–≤—ñ—á–∫–∏ –∑–∞–ø–∞–ª—é–≤–∞–Ω–Ω—è", "NGK", 120.00, 50), true);
    warehouse.addSupplyOrder(AutoPart("–ê–∫—É–º—É–ª—è—Ç–æ—Ä", "Varta", 3500.00, 5), true);
    warehouse.addSupplyOrder(AutoPart("–ì–∞–ª—å–º—ñ–≤–Ω—ñ –∫–æ–ª–æ–¥–∫–∏", "Brembo", 1800.00, 20), false);

    on_showAllButton_clicked();
}

void WarehouseWindow::updateStats(){
    ui->statsLabel->setText(QString(
        "üìä –°—Ç–∞—Ç–∏—Å—Ç–∏–∫–∞: –í—Å—å–æ–≥–æ –∑–∞–ø—á–∞—Å—Ç–∏–Ω: %1 | –£–Ω—ñ–∫–∞–ª—å–Ω–∏—Ö –Ω–∞–∑–≤: %2 | "
        "–ü–æ—Å—Ç–∞—á–∞–Ω—å –≤ —á–µ—Ä–∑—ñ: %3 | –Ü—Å—Ç–æ—Ä—ñ—è: %4 –∑–∞–ø–∏—Å—ñ–≤")
        .arg(warehouse.size())
        .arg(warehouse.getUniqueNamesCount())
        .arg(warehouse.getSupplyQueueSize())
        .arg(warehouse.hasHistory() ? "—î" : "–Ω–µ–º–∞—î"));
}

void WarehouseWindow::onTableCellDoubleClicked(int row, int column){
    Q_UNUSED(column);
    if (row < 0 || row >= ui->tableWidget->rowCount()) return;
    
    QString name = ui->tableWidget->item(row, 0)->text();
    QString manufacturer = ui->tableWidget->item(row, 1)->text();
    string key = name.toStdString() + "_" + manufacturer.toStdString();
    
    updatePart(key);
}
